index.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+// See https://github.com/jshint/jshint/issues/1747 for context
+/* global -Promise */
+var Promise = require('es6-promise').Promise;
+var adb = require('adbkit');
+var portfinder = require('portfinder');
+
+var REMOTE_DEBUGGER_SOCKET = 'localfilesystem:/data/local/debugger-socket';
+
+module.exports = forwardPorts;
+
+function forwardPorts(options) {
+  options = options || {};
+  var devices = options.devices;
+
+  var client = adb.createClient();
+
+  return new Promise(function(resolve, reject) {
+    if (devices === undefined || !Array.isArray(devices)) {
+      return reject(new Error('an array of devices is required to forward ports'));
+    }
+    return resolve();
+  }).then(function() {
+    return findExistingPortForwards(client, devices);
+  }).then(function(devicesWithPorts) {
+    return forwardDevices(client, devicesWithPorts);
+  });
+}
+
+function findExistingPortForwards(client, devices) {
+  return client.listForwards().then(function(ports) {
+
+    // Index the list of forwarded ports by device ID.
+    var portMap = {};
+    ports.forEach(function(port) {
+
+      // Filter for remote debugging sockets, just in case something else has
+      // been forwarded.
+      if (port.remote !== REMOTE_DEBUGGER_SOCKET) {
+        return;
+      }
+
+      // Add this port to a list associated with the device ID
+      if (!portMap[port.serial]) {
+        portMap[port.serial] = [];
+      }
+      portMap[port.serial].push(port);
+
+    });
+
+    // Annotate the list of devices with associated ports.
+    return devices.map(function(device) {
+      device.ports = portMap[device.id] || [];
+      return device;
+    });
+
+  });
+}
+
+function forwardDevices(client, devices) {
+  return Promise.all(devices.map(function(device) {
+
+    if (device.ports.length > 0) {
+      // Skip forwarding for this device, but ensure TCPIP ports are in a more
+      // friendly format.
+      device.ports.forEach(function(port) {
+        port.port = (port.local.indexOf('tcp:') === -1) ? null :
+          port.local.replace('tcp:', '');
```

Comment:
It also might be a case of “using an if statement is more lines and repetition but improves readability”. I actually love the ternary operator but it might not improve legibility here.